lnp2p/src/legacy/types.rs

```diff
@@ -744,6 +746,63 @@ impl LightningDecode for AddressList {
     }
 }
 
+#[derive(Display, Copy, Clone, Debug, PartialEq, Hash, Eq, PartialOrd, Ord, Wrapper)]
+#[display("{0} sec")]
+pub struct Timestamp(u32);
+
+impl<T> From<DateTime<T>> for Timestamp
+where
+    T: TimeZone,
+{
+    fn from(dt: DateTime<T>) -> Timestamp {
+        Timestamp(dt.timestamp() as u32)
+    }
+}
+
+impl From<u32> for Timestamp {
+    fn from(seconds: u32) -> Timestamp {
+        Timestamp(seconds)
+    }
+}
+
+impl Default for Timestamp {
+    fn default() -> Self {
+        Timestamp(Utc::now().timestamp() as u32)
+    }
+}
+
+impl Timestamp {
+    pub fn to_utc_datetime(&self) -> DateTime<Utc> {
+        Utc.timestamp(self.0 as i64, 0)
+    }
+
+    pub fn to_local_datetime(&self) -> DateTime<Local> {
+        Local.timestamp(self.0 as i64, 0)
+    }
+}
+
+impl Add for Timestamp {
+    type Output = Timestamp;
+
+    fn add(self, rhs: Timestamp) -> Timestamp {
```

Comment:
This and `sub` above should not be a `Add` trait implementation, since returning zero time on overflow is incorrect and a bug. We need to do just `impl Timestamp` with methods named `checked_sub` and `checked_add`, returning optional like underlying methods

Reply:
Do you mean something like this?

```rust
impl Timestamp{
    fn checked_add(self, rhs: Timestamp) -> Timestamp {
        self.0.checked_add(rhs.0)
    }

    fn checked_sub(self, rhs: Timestamp) -> Timestamp {
        self.0.checked_sub(rhs.0)
    }
}
```

Reply:
Yes, exactly

Reply:
Bit this should return `Option`s, otherwise compilation will report a error

Reply:
oh that's right sorry, fixing it

Reply:
take a look
```rust
impl Timestamp {
  pub fn checked_add(self, rhs: Timestamp) -> Option<Timestamp> {
      Timestamp(self.0.checked_add(rhs.0).unwrap()).into()
  }

  pub fn checked_sub(self, rhs: Timestamp) -> Option<Timestamp> {
      Timestamp(self.0.checked_sub(rhs.0).unwrap()).into()
  }
}
```

Reply:
That will not work and will panic on each overflow. Use `Option::map` pls

Reply:
like this?
```rust
impl Timestamp {
    pub fn checked_add(self, rhs: Timestamp) -> Option<Timestamp> {
        self.0.checked_add(rhs.0).map(Timestamp)
    }

    pub fn checked_sub(self, rhs: Timestamp) -> Option<Timestamp> {
        self.0.checked_sub(rhs.0).map(Timestamp)
    }
}
```